fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/PruneOlapScanPartition.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.doris.nereids.trees.expressions.LessThanEqual;
+import org.apache.doris.nereids.trees.expressions.Or;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.expressions.literal.Literal;
+import org.apache.doris.nereids.trees.plans.logical.LogicalFilter;
+import org.apache.doris.nereids.trees.plans.logical.LogicalOlapScan;
+import org.apache.doris.nereids.util.ExpressionUtils;
+import org.apache.doris.nereids.util.Utils;
+import org.apache.doris.planner.ColumnBound;
+import org.apache.doris.planner.ColumnRange;
+import org.apache.doris.planner.PartitionPruner;
+import org.apache.doris.planner.RangePartitionPrunerV2;
+import org.apache.doris.planner.ScanNode.ColumnRanges;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Range;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Used to prune partition of olap scan, should execute after SwapProjectAndFilter, MergeConsecutiveFilters,
+ * MergeConsecutiveProjects and all predicate push down related rules.
+ */
+public class PruneOlapScanPartition extends OneRewriteRuleFactory {
+
+    @Override
+    public Rule build() {
+        return logicalFilter(logicalOlapScan()).thenApply(ctx -> {
+            LogicalFilter<LogicalOlapScan> filter = ctx.root;
+            LogicalOlapScan scan = filter.child();
+            Expression predicate = filter.getPredicates();
+            List<Expression> expressionList = ExpressionUtils.extractConjunction(predicate);
+            OlapTable table = scan.getTable();
+            Set<String> partitionColumnNameSet = Utils.execWithReturnVal(table::getPartitionColumnNames);
+            PartitionInfo partitionInfo = table.getPartitionInfo();
+            // TODO: 1. support grammar: SELECT * FROM tbl PARTITION(p1,p2)
+            //       2. support list partition
+            if (partitionColumnNameSet.isEmpty() || !partitionInfo.getType().equals(PartitionType.RANGE)) {
+                return ctx.root;
+            }
+            // TODO: Process all partition column for now, better to process required column only.
+            Map<String, ColumnRange> columnNameToRange = Maps.newHashMap();
+            for (String colName : partitionColumnNameSet) {
+                ColumnRange columnRange = createColumnRange(colName, expressionList);
+                columnNameToRange.put(colName, columnRange);
+            }
+
+            Map<Long, PartitionItem> keyItemMap = partitionInfo.getIdToItem(false);
+            PartitionPruner partitionPruner = new RangePartitionPrunerV2(keyItemMap,
+                    partitionInfo.getPartitionColumns(), columnNameToRange);
+            Collection<Long> selectedPartitionId =  Utils.execWithReturnVal(partitionPruner::prune);
+            scan.getSelectedPartitionIds().retainAll(selectedPartitionId);
```

Comment:
REMEMBER THAT: `Plan` and `Expression` in Nereids is **_IMMUTABLE_**.
we should not mod scan's attribute. Instead, we need to generate a new `LogicalOlapScan` with pruned partitionIds.